mitiq/shadows/post_processing.py

```diff
@@ -0,0 +1,141 @@
+from typing import Tuple, List, Union, Any
+
+import numpy as np
+
+PAULI_MAP = {"X": 0, "Y": 1, "Z": 2}
+
+
+def snapshot_state(
+    b_list: List[Union[int, float]], u_list: List[str]
+) -> np.ndarray[Any, Any]:
+    """
+    Implement a single snapshot state reconstruction,
+
+    Args:
+        b_list (array): The list of classical outcomes for the snapshot.
+        u_list (array): Indices for the applied Pauli measurement.
+
+    Returns:
+        Numpy array with the reconstructed snapshot.
+    """
+    num_qubits = len(b_list)
+
+    # computational basis states, e.g.|0>=(1,0)
+    zero_state = np.array([[1, 0], [0, 0]])
+    one_state = np.array([[0, 0], [0, 1]])
+
+    # local unitary that applied to the computational basis states, e.g.
+    # $$Z-$$basis measurement,
+    # which equivalent to a random Pauli measurement, i.e. for each qubit,
+    # we randomly decide to measure the Pauli operators.
+    phase_z = np.array([[1, 0], [0, -1j]], dtype=np.complex128)
+    hadamard = np.array([[1, 1], [1, -1]]) / np.sqrt(2)
+    identity = np.eye(2)
+
+    # act $$Ad_{U^\dagger}$$ on the computational basis states, where $$U$$ is
+    # Define ensemble which the random unitary sample from.
+    unitaries = [hadamard, hadamard @ phase_z, identity]
+
+    # reconstructing a single snapshot state by applying Eq. (S44)
+    rho_snapshot = np.array([1], dtype=np.complex128)
+    for i in range(num_qubits):
+        state = zero_state if b_list[i] == 1 else one_state
+        U = unitaries[PAULI_MAP[u_list[i]]]
+
+        # applying Eq. (S44)
+        local_rho = 3 * (U.conj().T @ state @ U) - identity
+        rho_snapshot = np.kron(rho_snapshot, local_rho)
+
+    return rho_snapshot
+
+
+def shadow_state_reconstruction(
+    measurement_outcomes: Tuple[np.ndarray[Any, Any], np.ndarray[Any, Any]]
+) -> np.ndarray[Any, Any]:
+    """
+    Reconstruct a state approximation as an average over all snapshots.
+
+    Args:
+        measurement_outcomes (tuple): A shadow tuple obtained
+        from `z_basis_measurement`.
+
+    Returns:
+        Numpy array with the reconstructed quantum state.
+    """
+    num_snapshots, num_qubits = measurement_outcomes[0].shape
+
+    # classical values
+    b_lists, u_lists = measurement_outcomes
+
+    # Averaging over snapshot states.
+    shadow_rho = np.zeros(
+        (2**num_qubits, 2**num_qubits), dtype=np.complex128
+    )
+    for i in range(num_snapshots):
+        shadow_rho += snapshot_state(b_lists[i], u_lists[i])
+
+    return shadow_rho / num_snapshots
+
+
+def expectation_estimation_shadow(
+    measurement_outcomes: Tuple[np.ndarray[Any, Any], np.ndarray[Any, Any]],
+    observable: Any,
```

Comment:
Missing meaningful type. There is one in the docstring, but if we are looking to use an observable here, maybe we should try to use Mitiq's [observable](https://mitiq.readthedocs.io/en/stable/guide/observables.html) class.

Reply:
Does this work with Mitiq's observable class?

Reply:
Thanks! I have correct it, please check the new version!